#Scripts to plot manuscript figures

```{r setup}
source('../Recursive_Partitioning/ExtractSubsample.R')
source('../Load_Phenotype/Load_Phenotype.R')
source('../Gene_Phenotype_Association/GenePhenotypeAssociation.R')
source('../Load_Bgen/LoadBgen.R')

library(glue)
library(kableExtra)
library(corrplot)
library(qqman)
library(partykit)
library(ggplot2)
```

```{r hmgcr_cor}
#Fig A1

includedSNPS <- c('rs12916','rs17238484','rs5909','rs2303152','rs10066707','rs2006760')

path <-  '~/bsu_scratch/LDL_Project_Data_Aug2019/Genotype_Data/'
bgen_file_prefix <- 'ukb_imp_chr#_HRConly'

dosageMatrix <- LoadBgen(path,bgen_file_prefix,includedSNPS)
r2_matrix <- matrix(nrow = length(includedSNPS),ncol = length(includedSNPS))
row.names(r2_matrix) <- includedSNPS
colnames(r2_matrix) <- includedSNPS

for(i in 1:nrow(r2_matrix)){
  for(j in i:ncol(r2_matrix)){
    r2_matrix[i,j] <- cor(as.vector(dosageMatrix[i,]),as.vector(dosageMatrix[j,])) ^ 2
  }
}
corrplot(r2_matrix,type='upper',method = 'color',addCoef.col = 'black')

```

```{r hmgcr_score}
#Table A1

#Get effect sizes for known snps
known_snps = c('rs12916','rs17238484','rs5909','rs2303152','rs10066707','rs2006760')
target_LDL_known <- CalcSnpPhenoAssociation(path = '~/bsu_scratch/LDL_Project_Data/Genotype_Data/',
                                            sample_file_prefix = 'ukbb_LDL_metadata_with_PC',
                                            bgen_file_prefix = 'ukb_imp_chr#_HRConly',
                                            phenotype = 'LDLdirect',
                                            known_snps,1,cov='sex,ages,bmi',PC=5,med=1,16,F)
target_LDL_known <- dplyr::left_join(target_LDL_known,AllSNPsOfGene('HMGCR',100000,100000),by = c('rsid'='rsid'))

```

```{r baseline_characteristics}
#Table 1

#Load Phenotype
raw_pheno <- data.table::fread('~/bsu_scratch/LDL_Project_Data/Genotype_Data/ukbb_LDL_metadata_with_PC.csv')
pheno_file <- LoadPhenotype(path = '~/bsu_scratch/LDL_Project_Data_Aug2019/Genotype_Data/',
                            sample_file_prefix = 'ukbb_LDL_metadata_with_PC',
                            phenotype = 'LDLdirect',
                            cov_names = c('sex','ages','bmi'),
                            eur_only = 1,
                            med = 1)
dbp <- raw_pheno$dbp[pheno_file$samplesToKeep]
sbp <- raw_pheno$sbp[pheno_file$samplesToKeep]
#Load HMGCR gene score
hmgcr_score <- readRDS('~/bsu_scratch/LDL_Project_Data_Aug2019/Random_Forest/rs12916_rs17238484_rs5909_rs2303152_rs10066707_rs2006760_LDLdirect/data_p_9e-06_maf_5e-2.rds')$dosageTarget
hmgcr_score_med <- median(hmgcr_score)

age_overall <- round(mean(pheno_file$covariates$ages),2)
age_overall_sd <- round(sd(pheno_file$covariates$ages),2)
age_below_med <- round(mean(pheno_file$covariates$ages[hmgcr_score <= hmgcr_score_med]),2)
age_below_med_sd <- round(sd(pheno_file$covariates$ages[hmgcr_score <= hmgcr_score_med]),2)
age_above_med <- round(mean(pheno_file$covariates$ages[hmgcr_score > hmgcr_score_med]),2)
age_above_med_sd <- round(sd(pheno_file$covariates$ages[hmgcr_score > hmgcr_score_med]),2)

bmi_overall <- round(mean(pheno_file$covariates$bmi),2)
bmi_overall_sd <- round(sd(pheno_file$covariates$bmi),2)
bmi_below_med <- round(mean(pheno_file$covariates$bmi[hmgcr_score <= hmgcr_score_med]),2)
bmi_below_med_sd <- round(sd(pheno_file$covariates$bmi[hmgcr_score <= hmgcr_score_med]),2)
bmi_above_med <- round(mean(pheno_file$covariates$bmi[hmgcr_score > hmgcr_score_med]),2)
bmi_above_med_sd <- round(sd(pheno_file$covariates$bmi[hmgcr_score > hmgcr_score_med]),2)

male_overall <- round(sum(pheno_file$covariates$sex)/length(pheno_file$covariates$sex) * 100,2)
male_below_med <- round(sum(pheno_file$covariates$sex[hmgcr_score <= hmgcr_score_med]) / sum(hmgcr_score <= hmgcr_score_med) * 100,2)
male_above_med <- round(sum(pheno_file$covariates$sex[hmgcr_score > hmgcr_score_med]) / sum(hmgcr_score > hmgcr_score_med) * 100,2)

sbp_overall <- round(mean(sbp,na.rm = T),2)
sbp_overall_sd <- round(sd(sbp,na.rm = T),2)
sbp_below_med <- round(mean(sbp[hmgcr_score <= hmgcr_score_med],na.rm = T),2)
sbp_below_med_sd <- round(sd(sbp[hmgcr_score <= hmgcr_score_med],na.rm = T),2)
sbp_above_med <- round(mean(sbp[hmgcr_score > hmgcr_score_med],na.rm = T),2)
sbp_above_med_sd <- round(sd(sbp[hmgcr_score > hmgcr_score_med],na.rm = T),2)


dbp_overall <- round(mean(dbp,na.rm = T),2)
dbp_overall_sd <- round(sd(dbp,na.rm = T),2)
dbp_below_med <- round(mean(dbp[hmgcr_score <= hmgcr_score_med],na.rm = T),2)
dbp_below_med_sd <- round(sd(dbp[hmgcr_score <= hmgcr_score_med],na.rm = T),2)
dbp_above_med <- round(mean(dbp[hmgcr_score > hmgcr_score_med],na.rm = T),2)
dbp_above_med_sd <- round(sd(dbp[hmgcr_score > hmgcr_score_med],na.rm = T),2)

lipd_overall <- round(sum(pheno_file$lipdbin == 'Current')/length(pheno_file$lipdbin) * 100,2)
lipd_below_med <- round(sum(pheno_file$lipdbin[hmgcr_score <= hmgcr_score_med] == 'Current') / sum(hmgcr_score <= hmgcr_score_med) * 100,2)
lipd_above_med <- round(sum(pheno_file$lipdbin[hmgcr_score > hmgcr_score_med] == 'Current') / sum(hmgcr_score > hmgcr_score_med) * 100,2)

df_merged <- data.frame(Overall=c(glue("{age_overall} ({age_overall_sd})"),
                                  glue("{bmi_overall} ({bmi_overall_sd})"),
                                  glue("{male_overall}%"),
                                  glue("{sbp_overall} ({sbp_overall_sd})"),
                                  glue("{dbp_overall} ({dbp_overall_sd})"),
                                  glue("{round(mean(pheno_file$phenotypes),3)} ({round(sd(pheno_file$phenotypes),3)})"),
                                  glue("{lipd_overall}%")),
                        Below_Med=c(glue("{age_below_med} ({age_below_med_sd})"),
                                  glue("{bmi_below_med} ({bmi_below_med_sd})"),
                                  glue("{male_below_med}%"),
                                  glue("{sbp_below_med} ({sbp_below_med_sd})"),
                                  glue("{dbp_below_med} ({dbp_below_med_sd})"),
                                  glue("{round(mean(pheno_file$phenotypes[hmgcr_score <= hmgcr_score_med]),3)} ({round(sd(pheno_file$phenotypes[hmgcr_score <= hmgcr_score_med]),3)})"),
                                  glue("{lipd_below_med}%")),
                        Above_Med=c(glue("{age_above_med} ({age_above_med_sd})"),
                                  glue("{bmi_above_med} ({bmi_above_med_sd})"),
                                  glue("{male_above_med}%"),
                                  glue("{sbp_above_med} ({sbp_above_med_sd})"),
                                  glue("{dbp_above_med} ({dbp_above_med_sd})"),
                                  glue("{round(mean(pheno_file$phenotypes[hmgcr_score > hmgcr_score_med]),3)} ({round(sd(pheno_file$phenotypes[hmgcr_score > hmgcr_score_med]),3)})"),
                                  glue("{lipd_above_med}%"))
)
colnames(df_merged) <- c(glue("Overall\n(N={length(hmgcr_score)})"),
                         glue("HMGCR Score <= Median \n(N={sum(hmgcr_score <= hmgcr_score_med)})"),
                         glue("HMGCR Score > Median \n(N={sum(hmgcr_score > hmgcr_score_med)})"))
rownames(df_merged) <- c("Age,years",
                         "Body Mass Index,kg/m2",
                         "Male,%",
                         "Systolic Blood Pressure,mmHg",
                         "Diastolic Blood Pressure,mmHg",
                         "LDL Cholestrol,mmol/L",
                         "Current Use of Cholestrol Lowering Medication,%")
age_p <- t.test(pheno_file$covariates$ages[hmgcr_score <= hmgcr_score_med],pheno_file$covariates$ages[hmgcr_score > hmgcr_score_med])$p.value
bmi_p <- t.test(pheno_file$covariates$bmi[hmgcr_score <= hmgcr_score_med],pheno_file$covariates$bmi[hmgcr_score > hmgcr_score_med])$p.value
male_tbl <- cbind(table(pheno_file$covariates$sex[hmgcr_score <= hmgcr_score_med]),table(pheno_file$covariates$sex[hmgcr_score > hmgcr_score_med]))
male_p <- chisq.test(male_tbl,correct = F)$p.value
# male_p <-  prop.test(x = c(sum(pheno_file$covariates$sex[hmgcr_score <= hmgcr_score_med] == 1),sum(pheno_file$covariates$sex[hmgcr_score > hmgcr_score_med] == 1)),n=c(sum(hmgcr_score <= hmgcr_score_med),sum(hmgcr_score > hmgcr_score_med)))$p.value
sbp_p <- t.test(sbp[hmgcr_score <= hmgcr_score_med],sbp[hmgcr_score > hmgcr_score_med])$p.value
dbp_p <- t.test(dbp[hmgcr_score <= hmgcr_score_med],dbp[hmgcr_score > hmgcr_score_med])$p.value
ldl_p <- t.test(pheno_file$phenotypes[hmgcr_score <= hmgcr_score_med],pheno_file$phenotypes[hmgcr_score > hmgcr_score_med])$p.value

lipdbin <- pheno_file$lipdbin
lipdbin[lipdbin=='']<- 'Other'
lipd_tbl <- cbind(table(lipdbin[hmgcr_score <= hmgcr_score_med]),table(lipdbin[hmgcr_score > hmgcr_score_med]))
lipd_p <- chisq.test(lipd_tbl)$p.value
# lipd_p <- prop.test(x = c(sum(pheno_file$lipdbin[hmgcr_score <= hmgcr_score_med] == 'Current'),sum(pheno_file$lipdbin[hmgcr_score > hmgcr_score_med] == 'Current')),n=c(sum(hmgcr_score <= hmgcr_score_med),sum(hmgcr_score > hmgcr_score_med)))$p.value

```

```{r manhattan}
# Fig A3

pos_info <- data.table::fread('~/bsu_scratch/UKB_Data/ukb_imp_allchr_HRConly_samplefilt_snpstats.txt',select = c('rsid','chromosome','position','minor_allele_frequency'))
interaction_data <- data.table::fread('~/bsu_scratch/LDL_Project_Data_Aug2019/Interaction_Data/rs12916_rs17238484_rs5909_rs2303152_rs10066707_rs2006760_LDLdirect_known/allchr.txt',select = c('rsid','coeff_target','p_target','coeff_snp','p_snp','coeff_int','p_int'),header = T)
interaction_data <- dplyr::left_join(interaction_data,pos_info,by=c('rsid'='rsid')) 
#Exlude snps 20Mb upstream and downstream from HMGCR

chr <- 5
start_pos <- 74632154 - 2e6
end_pos <- 74657929 + 2e6
interaction_data_excl_hmgcr <- interaction_data %>% dplyr::filter(!(chromosome == chr & position > start_pos & position < end_pos))

png('interaction_manhattan.png',width = 1600,height = 600)
qqman::manhattan(interaction_data_excl_hmgcr,chr='chromosome',bp = 'position',p = 'p_int',snp = 'rsid',suggestiveline = F,cex.axis = 0.8,cex.lab = 1.3,ylab='')
title(ylab=expression('-log'[10]*'(p)'), mgp=c(1.5,1.5,0),cex.lab=1.3)
dev.off()
knitr::include_graphics('interaction_manhattan.png')

```


```{r qq}
#Fig A4

z = qnorm(interaction_data_excl_hmgcr$p_int / 2)
lambda = round(median(z^2) / qchisq(0.5, 1),6)
print(lambda)

png('qq_plot.png',width = 600,height = 600)
qqman::qq(pvector = interaction_data_excl_hmgcr$p_int)
text(1,5,labels = bquote(lambda == .(lambda)))
dev.off()
knitr::include_graphics('qq_plot.png')
```

```{r example_tree}
#Fig 1

library(partykit)
example_tree <- readRDS('~/bsu_scratch/LDL_Project_Data_Aug2019/Random_Forest_Old/rs12916_rs17238484_rs5909_rs2303152_rs10066707_rs2006760_LDLdirect/0.75_30000_7e-6_5e-2/tree18.rds')

pred <- function(i) {
  beta_split <- strsplit(x=unlist(i),split = ' ')[[1]][1]
  beta_string <- paste0('Effect =',beta_split)
  n_split <- strsplit(x=unlist(i),split = ' ')[[1]][3]
  n_string <- paste0('\nN = ',n_split)
  paste0(beta_string,n_string)
}
pdf(file = '~/MRC_BSU_Internship_LDL/Final_Figures/exampled_tree.pdf',width = 8,height = 5)
plot(example_tree$bootstrapPartyTree,terminal_panel = node_terminal,tp_args = list(FUN = pred),pop = F)
dev.off()
```

```{r sd_plot}
#Fig 5
#Fig A6

source('../Weighted_SD/CompareTreeSD.R')

ggarrange(p4)
ggarrange(p2,p3)
```

```{r var_imp}
# Fig A5
var_imp_testing <- do.call(rbind,lapply(1:2000,function(i) readRDS(paste0('~/bsu_scratch/LDL_Project_Data_Aug2019/Random_Forest_Old/rs12916_rs17238484_rs5909_rs2303152_rs10066707_rs2006760_LDLdirect/0.75_30000_7e-6_5e-2/var_imp/var_imp_tree',i,'.rds'))$testingVarImp))
var_imp_training <- do.call(rbind,lapply(1:2000,function(i) readRDS(paste0('~/bsu_scratch/LDL_Project_Data_Aug2019/Random_Forest_Old/rs12916_rs17238484_rs5909_rs2303152_rs10066707_rs2006760_LDLdirect/0.75_30000_7e-6_5e-2/var_imp/var_imp_tree',i,'.rds'))$trainingVarImp))


df <- data.frame(SNP = c(names(colMeans(var_imp_training,na.rm = T)),names(colMeans(var_imp_testing,na.rm = T))),
                 Variable_Importance = c(colMeans(var_imp_training,na.rm = T),colMeans(var_imp_testing,na.rm = T)),
                 Data_Set = c(rep('Training',ncol(var_imp_training)),rep('Testing',ncol(var_imp_testing))))
var_imp_plot <- ggplot(data = df,
       aes(x = SNP, y = Variable_Importance,fill = Data_Set))+
  geom_bar(stat = "identity" ,position = 'dodge')+
  coord_flip() + ylab('Variable Importance') + labs(fill = 'Data Set')
ggpubr::ggexport(var_imp_plot,filename = '~/MRC_BSU_Internship_LDL/Final_Figures/var_imp.pdf')
```

```{r plot_hetero}
# Fig 4
source('PlotHetero.R')

p
```
